torch/lib/THD/base/data_channels/Store.cpp

```diff
@@ -0,0 +1,178 @@
+  }
+  // after sending the query, wait for a 'stop_waiting' reponse
+  query_type qr;
+  recv_bytes<query_type>(_socket, &qr, 1);
+  if (qr != query_type::stop_waiting)
+    throw std::runtime_error("stop_waiting response expected");
+}
+
+static void store_thread_daemon() {
+  port_type port;
+  rank_type world_size;
+  int socket;
+  std::tie(port, world_size) = load_master_env();
+  std::vector<int> sockets(world_size);
+  std::tie(socket, std::ignore) = listen(port);
+  // accept WORLD_SIZE connections
+  for (int i = 0; i < world_size; i++) {
+    int p_socket;
+    rank_type p_rank;
+    std::tie(p_socket, std::ignore) = accept(socket);
+    recv_bytes<rank_type>(p_socket, &p_rank, 1);
+    sockets[p_rank] = p_socket;
+  }
+
+  // listen for requests
+  struct pollfd fds[world_size];
+  for (int i = 0; i < world_size; i++) {
+    fds[i].fd = sockets[i];
+    fds[i].events = POLLIN;   // we only read
+  }
+
+  // receive the queries
+  bool finished = false;
+  while (!finished) {
+    for (int i = 0; i < world_size; i++) fds[i].revents = 0;
+    SYSCHECK(::poll(fds, world_size, -1));
+    for (int i = 0; i < world_size; i++) {
+      if (fds[i].revents == 0)
+        continue;
+      if (fds[i].revents ^ POLLIN)
+        throw std::system_error(ECONNABORTED, std::system_category());
+      finished = query(fds[i].fd);
+      if (finished)
+        break;
+    }
+  }
+}
+
+/* 
+ * query communicates with the worker. The format
+ * of the query is as follows:
+ * type of query | size of arg1 | arg1 | size of arg2 | arg2 | ...
+ * or, in the case of wait
+ * type of query | number of args | size of arg1 | arg1 | ...
+ */
+static bool query(int socket) {
+  static store_type store_;
+  static std::unordered_map<int, std::vector<std::string>> waiting;
+  query_type qt;
+  recv_bytes<query_type>(socket, &qt, 1);
+  if (qt == query_type::set) {
+    std::string key = recv_string(socket);
+    store_[key] = recv_vector<char>(socket);
+    // On "set" wake up all of the processes that wait
+    // for keys already in the store
+    for (auto& process : waiting) {
+      auto& p_socket = process.first;
+      auto& p_keys = process.second;
+      if (check(store_, p_keys)) {
+        wake_up(p_socket);
+        waiting.erase(p_socket);
+      }
+    }
+    return false;
+  } else if (qt == query_type::get) {
+    std::string key = recv_string(socket);
+    std::vector<char> data = store_.at(key);
+    send_vector(socket, data);
+    return false;
+  } else if (qt == query_type::wait) {
+    size_type nargs;
+    recv_bytes<size_type>(socket, &nargs, 1);
+    std::vector<std::string> keys(nargs);
+    for (int i = 0; i < nargs; i++) {
+      keys[i] = recv_string(socket);
+    }
+    if (check(store_, keys)) {
+      wake_up(socket);
+    } else {
+      waiting[socket] = std::move(keys);
+    }
+    return false;
+  } else if (qt == query_type::finish) {
+    return true;
+  } else {
+    throw std::runtime_error("expected a query type");
```

Comment:
Exactly! I thought that it might be handy to see that rubbish if something goes wrong.